akvo/rsr/usecases/change_project_parent.py

```diff
@@ -0,0 +1,144 @@
+
+
+def make_target_parent_map(tree, project_attr, object_id, target_id):
+    target_parent_map = {}
+    for node in tree:
+        if len(node['children']) < 2:
+            continue
+        children = node['children']
+        first, second = get_leave(children[0]), get_leave(children[1])
+        if first[project_attr] == object_id and second[project_attr] == target_id:
+            target_parent_map[first['id']] = second['id']
+        elif second[project_attr] == object_id and first[project_attr] == target_id:
+            target_parent_map[second['id']] = first['id']
+    return target_parent_map
+
+
+RF_MODELS_CONFIG = {
+    # key: (Model, parent_attribute, project_relation)
+    'results': (Result, 'parent_result', 'project'),
+    'indicators': (Indicator, 'parent_indicator', 'result__project'),
+    'periods': (IndicatorPeriod, 'parent_period', 'indicator__result__project'),
+    'dimension_names': (IndicatorDimensionName, 'parent_dimension_name', 'project'),
+    'dimension_values': (IndicatorDimensionValue, 'parent_dimension_value', 'name__project'),
+    'default_periods': (DefaultPeriod, 'parent_id', 'project'),
+}
+
+
+def get_rf_change_candidates(project, new_parent):
+    project_lineage_ids = [*get_lineage(project).values_list('id', flat=True)]
+    new_parent_lineage_ids = [*get_lineage(new_parent).values_list('id', flat=True)]
+    # no need to include project to find commond ancestor
+    common_ancestor_id = get_nearest_common_ancestor(project_lineage_ids[1:], new_parent_lineage_ids)
+    if common_ancestor_id is None:
+        print('No common ancestor found')
+        return {}
+    project_ids = set(project_lineage_ids).symmetric_difference(new_parent_lineage_ids) | {common_ancestor_id}
+    candidates = {}
+    for key, config in RF_MODELS_CONFIG.items():
+        model, parent_attr, project_relation = config
+        filter_arg = {f"{project_relation}__in": project_ids}
+        items = model.objects.filter(**filter_arg).values('id', parent_attr, project_relation)
+        items_tree = make_tree_from_list(items, parent_attr)
+        candidates[key] = make_target_parent_map(items_tree, project_relation, project.id, new_parent.id)
+    return candidates
+
+
+def change_parent(project, new_parent, verbosity=0):
+    old_parent = project.parents_all().first()
+    if old_parent.id == new_parent.id:
+        if verbosity > 0:
+            print("New parent same as current parent")
+        return
+    # change parents of RF items
+    change_candidates = get_rf_change_candidates(project, new_parent)
+    for key, candidates in change_candidates.items():
+        model, parent_attr, _ = RF_MODELS_CONFIG[key]
+        items = model.objects.filter(id__in=candidates.keys())
+        for item in items:
+            if verbosity > 1:
+                print(f"Change {key} parent of {item} to {candidates[item.id]}")
+            setattr(item, f"{parent_attr}_id", candidates[item.id])
+            item.save()
```

Comment:
Would it be better to use queryset `update` to instead of calling `item.save`? The models have complicated `save` methods that try to update children on save, etc. We could short-circuit that and avoid running all of that code, if we are manually taking care of the order, etc. 

Something like this piece of untested code below:

```python
    for key, candidates in change_candidates.items():
        model, parent_attr, _ = RF_MODELS_CONFIG[key]
        for item_id, target_id in candidates.items():
            if verbosity > 1:
                print(f"Change {key} parent of {item_id} to {target_id}")
            model.objects.filter(id__in=[item_id]).update(**{f"{parent_attr}_id": target_id})
```